### ISSUE TITLE
Generic implementation of CRD upgrade preflight check

### ISSUE THREAD

**Issue by everettraven** [operator-framework/operator-controller#661]
This focuses on creating the initial CRD Upgrade Safety preflight check. This should be a minimal implementation focused on:
- Adding the preflight check to the preflight check registry
- Creating an extensible interface for adding checks to be run against CustomResourceDefinition resources

The work done to resolve this issue will be a foundation that allows for future work to be done in parallel and iteratively add functionality to this preflight check.

If the [proposal](https://github.com/carvel-dev/carvel/pull/729/files) for adding this functionality to [carvel-dev/kapp](https://github.com/carvel-dev/kapp) has been accepted prior to starting this work, all changes should be made against  [carvel-dev/kapp](https://github.com/carvel-dev/kapp), otherwise the changes should be made against https://github.com/everettraven/kapp/tree/feature/crd-upgrade-safety-preflight

**Comment by everettraven** [comment]
Done in carvel-dev/kapp as of https://github.com/carvel-dev/kapp/pull/901 - closing
